### ISSUE TITLE
Still getting "The application failed to initialize properly (0xc0000005)" error on some systems

### ISSUE THREAD

**Issue by lordmulder** [upx/upx#83]
## What's the problem (or question)?
My application _reproducibly_ fails to start with error message "The application failed to initialize properly (0xc0000005)" in some system after it was compressed by UPX v3.93. It also happens with UPX v3.94 (git-e5865f). But, interestingly, it does **not** happen with older UPX v3.91 at all.

**Important:** Thought that this had been fixed in UPX v3.93, after I first encountered it with UPX v3.92. But it has been brought to my attention that it _still_ happens for _some_ users. Indeed, I was able to reliably reproduce it on my "business" laptop (Windows 7 Pro) as well as in my Windows XP x64 VM. It does **not** happen on my main development machine. Also does **not** happen on my "private" laptop (Windows 10).

So, clearly, **not** all systems are affected. But those that _are_ affected seem to be effected reproducibly!

#### Screenshot:

![0xc0000005](https://cloud.githubusercontent.com/assets/470590/24677358/dc355970-1986-11e7-84c0-5ea530b66b53.png)

## What should have happened?
My application should have started, rather than failing to start :smirk:

## Do you have an idea for a solution?
Nope, but it must be a post UPX v3.91 regression, because UPX v3.91 works fine on **all** my machines, including those that get "The application failed to initialize properly (0xc0000005)" with UPX v3.93.

## How can we reproduce the issue?
I provide here my application that can be used to reproduce the bug:

**Original/Uncompressed  (works fine):**
https://sourceforge.net/projects/muldersoft/files/Simple%20x264%20Launcher/Experimental/x264_launcher.2017-04-01.original.zip/download

**Compressed by UPX v3.91 (works fine):**
https://sourceforge.net/projects/muldersoft/files/Simple%20x264%20Launcher/Experimental/x264_launcher.2017-04-01.upx-v391.zip/download

**Compressed by UPX v3.93 (crashes):**
https://sourceforge.net/projects/muldersoft/files/Simple%20x264%20Launcher/Experimental/x264_launcher.2017-04-01.upx-v393.zip/download

(Please note: Files that get compressed by UPX are "x264_launcher.exe", "MUtils32-1.dll" as well as "Qt*.dll". All other files **never** are compressed)

## Please tell us details about your environment.

* UPX version used (`upx --version`): "upx 3.93" + "upx 3.94-git-e5865f171a5f"
* Operating System and version: Windows 7 Pro (64-Bit) + Windows XP x64 SP-2
* CPU architecture: Core i5-4300U / Core i7-6700K

**Comment by ml1050** [comment]
Unfortunately I can not reproduce the problem with wine+linux.

**Comment by lordmulder** [comment]
Well, I only can reproduce it on **two** of my machines as well.

That's why I first though v3.93 has fixed the problem that appeared in v3.92. But apparently it didn't.

Here I show you the problem with v3.93:
http://muldersoft.com/temp/GVbsVBSXhCT3woAD/upx-bug-v393.mp4.zip

For comparison, no problem with v3.91:
http://muldersoft.com/temp/GVbsVBSXhCT3woAD/upx-bug-v391.mp4.zip

**Comment by lordmulder** [comment]
Just to let you know: Problem still persists with new version 3.94 :unamused:

**Comment by Mattiwatti** [comment]
I don't have a fix for this issue, but I did some experimenting and this may help narrow it down:
- I could only reproduce in Windows XP x64.
- The guilty file is MUtils32-1.dll. You can check this by taking the non-compressed directory and putting a UPX > 3.91 compressed version of the file in it (crash reproduces), or the inverse to get a working compressed directory.
- The segfault happens in the ntdll loader during relocation processing. This means there is an issue with the image file (presumably relocations), not the stub code.
```
ntdll!LdrProcessRelocationBlockLongLong+36
ntdll!LdrRelocateImageWithBias+8F
ntdll!LdrRelocateImage+10
ntdll!LdrpMapDll+809
```

**Comment by GR-C** [comment]
Hello,
I've encountered the same problem with my application using upx 3.94. I can reproduce it on Windows 10 1709 by enabling the Windows Defender Exploit Guard Setting `Force randomization for images (Mandatory ASLR)` (Start `Windows Defender Security Center` > `App & browser control` (on the sidebar) > `Exploit protection` > `Exploit protection settings`). This setting is available since Windows 10 1709.

Compressing the app with upx version 3.91 does not lead to the problem.

To me (I am not very familiar in this topic) it looks like there is a mistake in the `BASE RELOCATIONS` section of the PE file. I figured it out using `dumpbin /Relocations`

**Comment by levicki** [comment]
@Mattiwatti @jreiser I am also experiencing this issue with my own 32-bit DLL packed with UPX 3.96. However, in my Windows 10 installation, Windows Defender is disabled using GPO.

I have tried disabling relocation stripping and export packing and it still crashes. However, if I pack the same DLL with UPX 3.91 it does not crash.

Do you have any idea what is causing this crash during relocation? There must have been some breaking change after 3.91 that causes this.

**Comment by levicki** [comment]
@markus-oberhumer @Mattiwatti @jreiser 

Let me add more info -- there are following differences when my DLL is packed with UPX 3.91 and 3.96:

|Section|UPX 3.91 Size|UPX 3.96 Size|
|-------|-------------:|-------------:|
|Import Directory|0x418|0x3F4|
|Base Relocation Directory|0x18|0x24|

Here are the contents of Base Relocation Directory as dumped with `dumpbin /Relocations`:

**UPX 3.91**
```
BASE RELOCATIONS #3
   17000 RVA,       18 SizeOfBlock
      BD  HIGHLOW            1000F000
     28C  HIGHLOW            100172A4
     290  HIGHLOW            100172AC
     294  HIGHLOW            10012804
     2E8  HIGHLOW            10012068
     2EC  HIGHLOW            1000FD00
     2F4  HIGHLOW            1000E258
       0  ABS                        
```
**UPX 3.96**
```
BASE RELOCATIONS #3
       0 RVA,        C SizeOfBlock
       4  HIGHLOW            00000000
       0  ABS                        
   17000 RVA,       18 SizeOfBlock
     22D  HIGHLOW            1000F000
     3FC  HIGHLOW            10017414
     400  HIGHLOW            1001741C
     404  HIGHLOW            10012804
     458  HIGHLOW            10012068
     45C  HIGHLOW            1000FD00
     464  HIGHLOW            1000E258
       0  ABS                        
```
I am not too familiar with relocation tables but as you can see, the one created with UPX 3.96 has one additional 0 entry at the start.

Let me know if there is any other information I can provide so this issue can be finally resolved.

**Comment by Mattiwatti** [comment]
Thanks @levicki. The second relocation table is invalid due to the HIGHLOW block with RVA 0 (RVA 0 is also used for ABS relocations as you can see, but this is normal). This will cause the ntdll loader to access an invalid address in `ntdll!LdrProcessRelocationBlockLongLong`. At least, this is the case for XP, where I can reproduce the original segfault. On Windows 7 the loader simply prevents the DLL from relocating. This means there is no crash, but also no image relocation. As for Windows 10, it's possible that the 'force ASLR' policy will do just that: force ASLR and prefer the segfault over a non-relocated image. But this is just a guess, I don't have a Windows 10 machine with this configuration so I can't test this.

I took another stab at trying to figure out what is causing the invalid relocation block to be added, and I found that the issue lies with the relocations for the TLS directory. Executables without TLS will not have this problem. However that alone is not enough (if it were, every DLL with a TLS directory would be corrupted by UPX which clearly is not the case - the issue only occurs with some files). I wish I could say more about the exact circumstances that cause this but at this point the UPX code started going over my head.

I did manage to fix the issue locally: after some debugging I found that the following code ([pefile.cpp#1380](https://github.com/upx/upx/blob/d7ba31cab8ce8d95d2c10e88d2ec787ac52005ef/src/pefile.cpp#L1380)) is what's adding the invalid relocation block:
```C
if __acc_cte(tls_handler_offset_reloc > 0) // tls_handler_offset_reloc = 4 for PE32, -1 for PE64
    rel->add(tls_handler_offset + tls_handler_offset_reloc, reloc_type); // tls_handler_offset = 0, reloc_type = 3 (IMAGE_REL_BASED_HIGHLOW)
```
I changed this to only add the relocation if `tls_handler_offset` is not 0:
```C
if __acc_cte(tls_handler_offset > 0 && tls_handler_offset_reloc > 0)
    rel->add(tls_handler_offset + tls_handler_offset_reloc, reloc_type);
```
With this change the issue is fixed for me, using the original DLL from the first post. I haven't made a PR for this because I'm not quite certain this is the correct fix (i.e. is is true that a `tls_handler_offset` of 0 is always invalid, and if so, is there not some underlying issue that should be properly fixed elsewhere instead?). But if you want you can compile this change and see if it fixes the problem for you.

**Comment by levicki** [comment]
@Mattiwatti You said:

> This will cause the ntdll loader to access an invalid address in `ntdll!LdrProcessRelocationBlockLongLong`.

This is exactly what is happening in Windows 10 1909 -- I was debugging this using WinDbg, and that's where it crashes. Moreover, the offset at which it crashes is always ending in 4 (i.e. `tls_handler_offset` is 0 and `tls_handler_offset_reloc` is 4) so we can be sure that the crash is caused by this invalid relocation entry.

> I changed this...

I looked at the code changes in pefile.cpp between UPX 3.91 and 3.92:

```
git diff 6766b14 1c01b77 src/pefile.cpp
```

I was trying to figure out where the protected class variable `tls_handler_offset` is being set, and I found it:

```
    if (use_tls_callbacks)
        tls_handler_offset = linker->getSymbolOffset("PETLSC2") + upxsection; 
```

Version 3.92 added `use_tls_callbacks`, and the associated functionality, which includes setting this value.

> I'm not quite certain this is the correct fix

If the TLS handler offset is 0 that probably means there is no TLS handler and there is no need for TLS callbacks?

I am not really very familiar with that part of PE specification, but this document seems to have a lot of information regarding TLS callbacks:
https://github.com/corkami/docs/blob/master/PE/PE.md

I hope UPX authors can chime in and tell us if there is a better way of fixing this.
